### PR TITLE
Implement Translatable in appropriate places

### PR DESCRIPTION
**Pull request by Machine-Maker**
Adding `Translatable` interface to a bunch of places

### CODE REVIEW

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch [hunk 1]
```diff
@@ -78,36 +108,79 @@ index 786b8011e98b2fe93cc2418d624f6350ede62d90..024deba760c41787190d20e4ac5c5419
...
 +    @Nullable
-+    String getTranslationKey() {
++    public String getTranslationKey() {
++        return org.bukkit.Bukkit.getUnsafe().getTranslationKey(this);
++    }
++
++    /**
++     * Return the translation key for the EntityType, so the client can translate it into the active
++     * locale when using a {@link net.kyori.adventure.text.TranslatableComponent}.<br>
++     * This is <code>null</code>, when the EntityType isn't known to NMS (custom entities)
```
Comment (Proximyst) [inline patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]:
```suggestion
+     * locale when using a {@link net.kyori.adventure.text.TranslatableComponent}.
+     * <p>
+     * This is {@code null} when the EntityType isn't known to NMS (custom entities)
```

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch [hunk 2]
```diff
@@ -118,9 +191,23 @@ index 9a878e4fde31c015e2f3fdf365d5d16c30198685..bd96c06aca1fc18807e65c34f399ce2e
...
 +     */
 +    @NotNull
++    @Deprecated
 +    public String getTranslationKey() {
++        return this.translationKey();
++    }
++
++    /**
++     * Gets the translation key for this itemstack.
++     * This is not the same as getting the translation key
```
Comment (Proximyst) [inline patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]:
```suggestion
+     * Gets the translation key for this itemstack.
+     * <p>
+     * This is not the same as getting the translation key
```

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch (hunk 1 above)
Comment (kezz) [reply patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]: The Javadoc is wrong here, you're throwing an exception not returning null (which would be correct as the API requires the return to not be null).

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch [hunk 3]
```diff
@@ -18,9 +27,21 @@ index 52290c43d1c02785c4cae4a73494a75cdc369e02..1efc97d88c38863bcd6cd4c11c8b88a1
...
 +     */
 +    @NotNull
++    @Deprecated
 +    public String getTranslationKey() {
++        return this.translationKey();
++    }
++
++    /**
++     * Return the translation key for the Material, so the client can translate it into the active
++     * locale when using a {@link net.kyori.adventure.text.TranslatableComponent}.
```
Comment (kezz) [inline patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]: Is there a reason why you're adding javadoc for these methods anyway?

**Review by kezz — APPROVED** [review]

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch (hunk 3 above)
Comment (Machine-Maker) [reply patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]: I guess it doesnt need one really. The doc on the interface probably suffices

**Comment by Machine-Maker** [comment]
Ok, if I missed any spots that this should be implemented on, let me know but I think I got everything.

**Comment by qixils** [comment]
well... depends on what counts as "should" 😅

here's some possible classes to add to, in varying amounts of sensibility:

- PotionEffectType (implemented in #5737)
- SoundCategory: `soundCategory.XXX`, where `XXX` is master, music, record, weather, hostile, neutral, player, block, ambient, voice
- Statistic: `stat.minecraft.XXX`, i'm not typing them out
- Biome, technically! `biome.minecraft.XXX`, XXX matches enum names i believe
- Difficulty: `options.difficulty.XXX`, XXX matches enum names
- ClientOption.ChatVisibility: `options.chat.visibility.XXX`, XXX matches enum names (except UNKNOWN ofc)
- ClientOption itself could technically use it but the translated text would be kinda awkward (i.e. "Chat" for visibility)
- FireworkEffect.Type: `item.minecraft.firework_star.shape.XXX`, XXX = small_ball, large_ball, star, creeper, burst
- GameMode? could be both `gameMode.XXX` which outputs like "Survival Mode" or `selectWorld.gameMode.XXX` which outputs like "Survival"
- UsageMode: `structure_block.mode.XXX`, XXX matches enum names
- MainHand: `options.mainHand.left` and `options.mainHand.right`
- Sound: for subtitles? probably wouldn't be feasible to implement, given a lot of sounds have no lang entry (and some sounds _should_ have a lang entry but don't and thus show up as untranslated keys)
- technically lots of legacy classes like TreeSpecies, GrassSpecies, CoalType, etc. but ew

didn't look in any event classes but that should be most/all of em that could maybe conceivably implement the interface

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch [hunk 4]
```diff
@@ -5,11 +5,45 @@ Subject: [PATCH] Add a way to get translation keys for blocks, entities and
  materials
```
Comment (qixils) [inline patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]: why rename the server patch but not the api patch?

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch [hunk 5]
```diff
@@ -78,36 +152,96 @@ index 786b8011e98b2fe93cc2418d624f6350ede62d90..024deba760c41787190d20e4ac5c5419
...
++     * @deprecated use {@link #translationKey()}
 +     */
++    @Deprecated
 +    @Nullable
-+    String getTranslationKey() {
++    public String getTranslationKey() {
++        return org.bukkit.Bukkit.getUnsafe().getTranslationKey(this);
++    }
++
++    /**
```
Comment (qixils) [inline patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]: should have a `{@inheritDoc}` i think

File: patches/server/0521-Add-methods-to-get-translation-keys.patch [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+     public net.minecraft.world.item.enchantment.Enchantment getHandle() {
+diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+index 6b79b1ea43b713105f37de517019c033477f9835..135749c789d44f0af1ad11ccd1b208f1e28853eb 100644
+--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
++++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+@@ -43,6 +43,7 @@ import org.bukkit.Registry;
+ import org.bukkit.UnsafeValues;
+ import org.bukkit.advancement.Advancement;
+ import org.bukkit.block.data.BlockData;
++import org.bukkit.craftbukkit.block.CraftBlock;
```
Comment (qixils) [inline patches/server/0521-Add-methods-to-get-translation-keys.patch]: accidental import?

**Comment by Machine-Maker** [comment]
> Statistic: stat.minecraft.XXX, i'm not typing them out

Ugly to implement with bukkit's bad Stat API, ill implement that through #5691 

> (all the others)

Ok, ill look to implement all or most of them.

File: patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch (hunk 5 above)
Comment (Machine-Maker) [reply patches/api/0225-Add-a-way-to-get-translation-keys-for-blocks-entitie.patch]: nah, not necessary, it auto inherits parts of the doc it doesnt have, like main description, return value, etc.

**Review by MiniDigger — APPROVED** [review]
requesting a look from @kezz and @kashike, but this looks good to me

**Review by kezz — APPROVED** [review]
LGTM! A few things missing but they can always be added later, no need to block this just because more stuff can implement it in the future.

**Review by jpenilla — APPROVED** [review]

**Pull request merged**